### PR TITLE
Restrict archive import source reads to configured downloader roots

### PR DESCRIPTION
**Pull request by Doezer**
# Pull Request _Template_

## Description

Follow-up to #1049, resolving the CodeQL `js/path-injection` limitation flagged in [this status comment](https://github.com/Doezer/Questarr/pull/1049#issuecomment-5720004531) on that PR. Stacked on top of `library-side-extraction-v2` (not `main`) because the flagged code — `walkRelative()`/`gatherFiles()` in `ImportStrategies.ts` — only exists on that still-open branch; `main` has an older, independently-rewritten version of this file with no `walkRelative` at all.

CodeQL's default path-injection query flagged 3 alerts: reads (`fs.stat`/`fs.readdir`) on a source path that ultimately derives from downloader-reported paths. The existing `isSensitivePath()` blocklist isn't a shape CodeQL recognizes as a sanitizer — it narrows what's forbidden but never proves the value is confined to a known-safe space, which is what CodeQL's taint-tracking wants to see before it clears the flow.

This adds that containment proof:
- `PathMappingService.getConfiguredRoots()` exposes the local filesystem roots the admin has actually mapped downloader paths onto — `translatePath()` already confines its own output to one of these internally, so this just surfaces that same trusted set.
- `PCImportStrategy` now takes that root list and enforces containment at both `planImport()` and `executeImport()` — two independent entry points, since `confirmImport` can hand `executeImport` a source path directly without ever going through `planImport` first.
- `ImportManager.resolveArchive()` gets the same check, since it reads the source directly before either of those runs.
- With **no mappings configured**, the check is a deliberate no-op: `translatePath()` already trusts the local filesystem wholesale in that case (common for single-host setups where the downloader and app share a filesystem), so there's no meaningful root set to restrict against — this doesn't change behavior for that common configuration.

**Confirmed against CI:** since this PR's base branch isn't `main`, most workflows (including CodeQL) don't trigger automatically on PRs/pushes here — the repo's Actions are scoped to `main`/`release/*`. I manually dispatched `codeql.yml` and `ci.yml` via `workflow_dispatch` against this branch to get a real signal rather than guessing. Result: **the `CodeQL` check now reports "No new alerts in code changed by this pull request"** — the exact same check that failed identically with "3 new alerts including 3 high severity security vulnerabilities" across every one of the 5 commits I pushed to #1049 trying earlier, less-targeted guards (a function-call blocklist, then an inline regex test). The allowlist-containment shape here — a real trusted-root set, `path.relative()` + `startsWith("..")`/`isAbsolute()`, enforced at every entry point before any read — is what actually clears it. All other checks (build, tests, lint, SonarCloud, SCA, secrets-scan, dependency-review) are green too.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue) — hardens source-path containment, resolves CodeQL alerts
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation — n/a, internal hardening of an existing check pattern
- [ ] If this PR adds a new actor/integration, external interface, or security-relevant change, I have updated docs/ARCHITECTURE.md, docs/API.md, and/or docs/SECURITY_ASSESSMENT.md accordingly — n/a, no new external interface; extends the existing `isSensitivePath` guard pattern
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes (`npx tsc --noEmit`, `npx vitest run`, `npm run lint` all clean — 2737 tests passing, 14 new) **and confirmed green on CI** via manual dispatch (see above)

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018PTsLFuBWFCMkfDhuXDWWj

## Summary by CodeRabbit

- **New Features**
  - Import operations now restrict source paths to configured downloader directories.
  - Path mappings resolve into a deduplicated list of configured local directories.
  - Import behavior remains unchanged when no directories are configured.

- **Bug Fixes**
  - Out-of-directory paths are rejected and flagged for manual review.
  - Improved protection against traversal, symlinks, and similarly named sibling directories.
  - Previews and import checks no longer access out-of-directory paths.

- **Tests**
  - Added coverage for containment checks, archive extraction, path mapping, symlinks, and unrestricted behavior.

### CODE REVIEW

File: server/services/ImportStrategies.ts [hunk 1]
```diff
@@ -368,6 +376,12 @@ export class PCImportStrategy implements ImportStrategy {
       throw new Error("Refusing to process a sensitive system path");
     }
 
+    assertWithinRoots(
+      sourcePath,
+      this.sourceRoots,
+      "Refusing to process a path outside the configured downloader roots"
+    );
```
[the remark this answers is not in the document]
Comment (Doezer) [reply server/services/ImportStrategies.ts]: Agreed this is real, but treating the full fix as out of scope here rather than pushing something half-done.

Genuinely closing this would mean holding a validated file descriptor (or directory handle) from the containment check through to the actual read/transfer/extraction, instead of re-resolving by pathname at each step. That's not practical with this pipeline as built: extraction shells out to 7z/unrar as external CLI processes that take a path argument, not a descriptor (there's no portable way to hand them `/proc/self/fd/N` cross-platform), and the move/copy/stat calls throughout go through fs-extra's path-based API, not Node's fd-based one. Doing this properly would mean swapping the filesystem interaction layer for the whole import pipeline — a much larger, separate change.

What I did push (in e43512d) narrows the window meaningfully: containment used to be checked once, far upstream, then trusted for the rest of the pipeline. Now `resolveArchive`, `planImport`, and `executeImport` each independently re-validate (with realpath, so a symlink swap is caught too) immediately before they read from the path, rather than relying on a single check done earlier. It doesn't eliminate the race the way fd-based validation would, but it's a real reduction given the realistic threat model here (an actor with write access to a configured downloader root, racing a filesystem swap on a self-hosted admin tool).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/path-security.ts [hunk 2]
```diff
@@ -15,3 +16,88 @@ export function isSensitivePath(rawPath: string): boolean {
...
+  // already-resolved pathname; there's nothing to canonicalize until it exists, and
+  // callers checking existence need that check to run regardless.
+  let canonicalCandidate: string;
+  try {
+    canonicalCandidate = await fs.realpath(verifiedCandidate);
+  } catch {
+    canonicalCandidate = verifiedCandidate;
+  }
+
+  const canonicalRoots = await Promise.all(resolvedRoots.map(canonicalizeRoot));
```
[the remark this answers is not in the document]
Comment (Doezer) [reply server/path-security.ts]: Confirmed and fixed in 35c22e3. The ENOENT fallback now canonicalizes just `matchedRoot` (which does exist) and re-appends the already-checked relative segment, instead of falling back to the fully lexical `verifiedCandidate`. Added a regression test with a real symlinked root and a not-yet-existing child.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Doezer** [comment]
Status: all checks are green as of bf17c6c, including CodeQL ("No new alerts in code changed by this pull request") — that took several iterations to get CodeQL to actually recognize the containment check as a sanitizer rather than a false positive; details in the commit messages for anyone curious about the specific guard shape it needed.

One more narrow gap worth documenting rather than chasing further: CodeRabbit's last pass (before hitting its review rate limit) noted that a symlink planted *inside* a configured root — as opposed to the root itself being a symlink, which is handled — combined with a not-yet-existing leaf under it, can still bypass full canonicalization on the ENOENT fallback path. Closing that fully would mean resolving the longest existing ancestor and appending the missing suffix, rather than canonicalizing just the top-level root.

I'm treating this as accepted under the same threat model already discussed on this PR for the TOCTOU finding: it requires an actor with write access to a configured downloader root to have planted a symlink there ahead of time, on a self-hosted admin tool. Given that, and that a full fix adds meaningful complexity for a case this narrow, I'm not pushing further changes for it now — flagging it here for visibility rather than leaving it unmentioned.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Doezer** [comment]
@coderabbitai review

File: server/path-security.ts [hunk 3]
```diff
@@ -15,3 +16,71 @@ export function isSensitivePath(rawPath: string): boolean {
...
+  // The realpath canonicalization and final containment check are nested directly
+  // inside the branch where a root's pathname check passes, rather than recording a
+  // match and continuing after the loop. CodeQL's path-injection sanitizer for the
+  // path.relative(root, x) + !startsWith("..") + !isAbsolute() shape only extends its
+  // "x is safe here" guarantee to code it can prove is dominated by that guarded
+  // branch; code reached after breaking out of a loop doesn't qualify, so fs.realpath
+  // needs to sit inside the same if as the check that clears it.
+  for (const root of resolvedRoots) {
+    const relative = path.relative(root, resolvedCandidate);
+    if (!relative.startsWith("..") && !path.isAbsolute(relative)) {
```
[the remark this answers is not in the document]
Comment (Doezer) [reply server/path-security.ts]: Fixed in bdf7a01, using `".." + path.sep` (string concatenation) rather than the suggested template literal — CodeQL's path-injection sanitizer recognizes that exact concatenation shape as equivalent to the literal `"../"` prefix check (verified against the upstream query source), while a template literal wouldn't match and would have reopened the alert this file went through several rounds to clear. Added regression tests for the false-rejection case and the exact-parent-of-root edge case.

---
_Generated by [Claude Code](https://claude.ai/code)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/path-security.ts [hunk 4]
```diff
@@ -15,3 +16,71 @@ export function isSensitivePath(rawPath: string): boolean {
...
+  }
+}
+
+/**
+ * Throws unless candidatePath resolves inside one of the given roots. An empty
+ * roots list means no restriction is configured — callers pass [] deliberately in
+ * that case (rather than skipping the call) to keep this the single place the
+ * containment logic lives.
+ */
+export async function assertWithinRoots(
```
[the remark this answers is not in the document]
Comment (Doezer) [reply server/path-security.ts:62]: Fixed in bdf7a01. Extracted the post-realpath canonical-roots check and the ENOENT fallback's canonicalization into their own functions — neither gates a filesystem call directly (unlike the pathname-containment check that immediately precedes `fs.realpath()`), so moving them out doesn't affect CodeQL's sanitizer recognition.

---
_Generated by [Claude Code](https://claude.ai/code)

---
_Generated by [Claude Code](https://claude.ai/code)_
